File: core/src/main/java/me/devsnox/planetsystem/core/listeners/PlanetListener.java
package me.devsnox.planetsystem.core.listeners;

import me.devsnox.planetsystem.api.holder.Holder;
import me.devsnox.planetsystem.api.player.PlanetPlayer;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.entity.Player;
import org.bukkit.event.*;
import org.bukkit.event.block.Action;
import org.bukkit.event.block.BlockBreakEvent;
import org.bukkit.event.block.BlockPlaceEvent;
import org.bukkit.event.player.PlayerInteractEvent;

public class PlanetListener implements Listener {

    @EventHandler(priority = EventPriority.LOWEST)
    public void onBlockPlace(final BlockPlaceEvent event) {
        this.blockBuild(event, event.getBlock(), event.getPlayer());
    }

    @EventHandler(priority = EventPriority.LOWEST)
    public void onBlockBreak(final BlockBreakEvent event) {
        this.blockBuild(event, event.getBlock(), event.getPlayer());
    }

    @EventHandler
    public void onPlayerInteract(PlayerInteractEvent event)  {
        if(Holder.Impl.holder.getWorld().equals(event.getPlayer().getWorld())) {
            if (event.getAction() == Action.PHYSICAL) {
                Block block = event.getClickedBlock();
                if (block == null) return;

                if (block.getType() == Material.SOIL) {
                    event.setUseInteractedBlock(Event.Result.DENY);
                    event.setCancelled(true);
                }
            }
        }
    }

    private void blockBuild(final Cancellable cancellable, final Block block, final Player player) {
        if (!Holder.Impl.holder.getWorld().equals(player.getWorld())) return;
        final PlanetPlayer planetPlayer = Holder.Impl.holder.getPlayerData().getPlayer(player.getUniqueId());
        if (planetPlayer != null && planetPlayer.canNotBuild(block)) cancellable.setCancelled(true);

        if (cancellable.isCancelled()) {
            planetPlayer.getLogger().warn("block_listener.deny_build");
        }
    }

}

File: core/src/main/kotlin/me/devsnox/planetsystem/core/listeners/PlanetListener.kt
package me.devsnox.planetsystem.core.listeners

import me.devsnox.planetsystem.api.holder.Holder.Impl.holder
import me.devsnox.planetsystem.api.holder.isNotInHolderWorld
import me.devsnox.planetsystem.core.log.MessageKeys
import net.darkdevelopers.darkbedrock.darkness.spigot.listener.Listener
import org.bukkit.Material
import org.bukkit.block.Block
import org.bukkit.entity.Player
import org.bukkit.event.Cancellable
import org.bukkit.event.Event
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.block.Action
import org.bukkit.event.block.BlockBreakEvent
import org.bukkit.event.block.BlockPlaceEvent
import org.bukkit.event.player.PlayerInteractEvent
import org.bukkit.plugin.java.JavaPlugin

class PlanetListener(javaPlugin: JavaPlugin) : Listener(javaPlugin) {

	@EventHandler(priority = EventPriority.LOW)
	fun onBlockPlaceEvent(event: BlockPlaceEvent) = blockBuild(event, event.block, event.player)

	@EventHandler(priority = EventPriority.LOW)
	fun onBlockBreakEvent(event: BlockBreakEvent) = blockBuild(event, event.block, event.player)

	@Suppress("SimplifyNegatedBinaryExpression")
	@EventHandler
	fun onPlayerInteractEvent(event: PlayerInteractEvent) {
		if (event.player.isNotInHolderWorld()) return
		if (event.action != Action.PHYSICAL) return
		if (!(event.clickedBlock?.type == Material.SOIL)) return

		cancel(event)
		event.setUseInteractedBlock(Event.Result.DENY)
	}

	private fun blockBuild(cancellable: Cancellable, block: Block, player: Player) {
		if (player.isNotInHolderWorld()) return

		val planetPlayer = holder.playerData.getPlayer(player.uniqueId) ?: return
		if (planetPlayer.canBuild(block)) return

		cancel(cancellable)
		planetPlayer.logger.warn(MessageKeys.LISTENER_PLANET_BUILD_DENY)
	}

}
